### PR TITLE
DrivAerML: lr=6e-4 + gc=1.0 + WD=1e-2 + T_max=10 — compound sweep

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The ongoing recipe transfer theme has students testing individual dimensions (gc, WD, T_max, LR). This PR tests a compound that fills a specific gap: lr=6e-4 (between 5e-4 baseline and 7e-4 faye is testing) combined with the full recipe compound gc=1.0+WD=1e-2+T_max=10. The motivation: faye tests lr=7e-4+gc=1.0 (no WD), wolfwood tests lr=8e-4+gc=1.0 (no WD), franky tests lr=4e-4+gc=1.0 (no WD). None test the compound gc+WD+LR shift simultaneously at lr=6e-4. AirfRANS golden LR is 7e-4 but DrivAerML may prefer a slightly smaller step.

## Instructions

**IMPORTANT: Use EXACTLY these run specifications.**

Run 1 — DrivAerML lr=6e-4 + gc=1.0 + WD=1e-2 + T_max=10:
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 6e-4 --cosine-t-max 10 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group edward-lr6e4-compound
```

Run 2 — DrivAerML lr=6e-4 + gc=1.0 + WD=1e-2 + T_max=30 (isolate LR effect with baseline T_max):
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 6e-4 --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group edward-lr6e4-compound
```

Run 3 — DrivAerML lr=6e-4 + gc=1.0 + T_max=10 (no WD, isolate scheduler+LR):
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 6e-4 --cosine-t-max 10 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --grad-clip 1.0 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group edward-lr6e4-compound
```

Run 4 — AirfRANS 2L/256d + gc=1.0 + WD=1e-2 + T_max=5 + lr=5e-4 (stabilization of kakashi's finding):
```bash
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 5e-4 --cosine-t-max 5 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 \
  --seed 44 \
  --wandb-group edward-lr6e4-compound
```

Report `val_primary/surface_rel_l2_pct` for DrivAerML and `val_primary/surface_mse` for AirfRANS.

## Baseline

- **DrivAerML val_primary/surface_rel_l2_pct:** 4.619% (PR #2691, epoch 256)
- **AirfRANS val_primary/surface_mse:** 0.001236 (PR #2828, epoch 349)
- **DrivAerML external target:** 3.71% (AB-UPT)
- **Reproduce golden DrivAerML:** `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### Summary

| Run | Config | Best `val_primary/surface_rel_l2_pct` | Best Epoch | Total Epochs | Status |
|-----|--------|--------------------------------------|------------|--------------|--------|
| 1 | DrivAerML lr=6e-4 + gc=1.0 + **WD=1e-2** + T_max=10 | 13.797% | 48 | 469 | ❌ Diverged (grad norms → ∞) |
| 2 | DrivAerML lr=6e-4 + gc=1.0 + **WD=1e-2** + T_max=30 | 6.041% | 144 | 474 | ❌ Diverged at ep194 |
| **3** | **DrivAerML lr=6e-4 + gc=1.0 + T_max=10 (no WD)** | **4.431%** | **451** | **484** | **✅ Beats baseline (4.619%)** |

| Run | Config | Best `val_primary/surface_mse` | Best Epoch | Total Epochs | Status |
|-----|--------|-------------------------------|------------|--------------|--------|
| 4 | AirfRANS 2L/256d + gc=1.0 + WD=1e-2 + T_max=5 + lr=5e-4 | 0.003349 | 177 | 748 (NaN from ep267) | ❌ Did not beat baseline (0.001236) |

### Baselines

- **DrivAerML baseline:** 4.619% (PR #2691, ep256) → **Run 3 achieves 4.431% (−4.1% relative improvement)**
- **DrivAerML external target:** 3.71% (AB-UPT) → Not reached
- **AirfRANS baseline:** 0.001236 (PR #2828, ep349) → Run 4 best 0.003349 (2.7× worse)

### Commands Used

**Run 1** (GPU 0):
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py --dataset drivaerml --optimizer adamw --lr 6e-4 --cosine-t-max 10 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --grad-clip 1.0 --weight-decay 1e-2 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group edward-lr6e4-compound --agent edward --wandb-name "edward/lr6e4-gc1-wd1e2-tmax10"
```

**Run 2** (GPU 1):
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py --dataset drivaerml --optimizer adamw --lr 6e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --grad-clip 1.0 --weight-decay 1e-2 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group edward-lr6e4-compound --agent edward --wandb-name "edward/lr6e4-gc1-wd1e2-tmax30"
```

**Run 3** (GPU 2):
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 python train.py --dataset drivaerml --optimizer adamw --lr 6e-4 --cosine-t-max 10 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --grad-clip 1.0 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group edward-lr6e4-compound --agent edward --wandb-name "edward/lr6e4-gc1-tmax10-noWD"
```

**Run 4** (GPU 3):
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=3 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 5e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --seed 44 --wandb-group edward-lr6e4-compound --agent edward --wandb-name "edward/airfrans-2L256d-gc1-wd1e2-tmax5"
```

### Peak Memory Usage

- DrivAerML runs (Runs 1-3): ~6.2 GiB per GPU
- AirfRANS run (Run 4): ~41.8 GiB

### W&B Run IDs

| Run | W&B ID | Link |
|-----|--------|------|
| 1 | `oqljsuy0` | [View](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/oqljsuy0) |
| 2 | `564acvay` | [View](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/564acvay) |
| 3 | `hwwgrf82` | [View](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/hwwgrf82) |
| 4 | `a5fcn7u7` | [View](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/a5fcn7u7) |

### What Happened

**Run 3 (lr=6e-4 + gc=1.0 + T_max=10, no WD) is the clear winner**, beating the DrivAerML baseline with **4.431% vs 4.619%**. Key observations:

1. **Weight decay (WD=1e-2) is destructive at lr=6e-4.** Both Runs 1 and 2 included WD and both diverged — Run 1 catastrophically from ep48 (grad norms → ∞), Run 2 more gradually (stable until ep144, diverged at ep194 with grad norms reaching 75+). The combination of lr=6e-4 with WD=1e-2 creates an overly aggressive effective regularization that destabilizes training once the model enters low-loss regimes.

2. **T_max=10 cosine cycling works well with lr=6e-4 + gc=1.0 when WD is absent.** Run 3 trained stably for all 484 epochs with grad norms staying around 0.08-0.11. The fast cosine restarts helped the model escape local minima repeatedly, with the metric oscillating in the 4.4-4.8% range in the final ~50 epochs.

3. **The ablation structure is informative:** Run 1 (T_max=10 + WD) vs Run 3 (T_max=10, no WD) isolates the weight decay effect. WD was the sole difference and caused complete divergence. Run 2 (T_max=30 + WD) showed that longer cosine periods delay but don't prevent the WD-induced instability.

4. **AirfRANS (Run 4) did not benefit from gc+WD+T_max=5.** Best of 0.003349 at ep177 is 2.7× worse than the 0.001236 baseline. The run went NaN at ep267, suggesting the compound recipe is not transferable to AirfRANS at these settings. The lr was 5e-4 (same as AirfRANS baseline), so the gc+WD+T_max=5 additions hurt rather than helped.

### Suggested Follow-ups

1. **New DrivAerML baseline candidate:** Run 3's recipe (lr=6e-4 + gc=1.0 + T_max=10, no WD, no EMA) at 4.431% should be tested with a longer training budget to see if it can push toward the 3.71% external target. The oscillation in the 4.4-4.8% range suggests it hasn't fully converged.

2. **Try lr=6e-4 + gc=1.0 + T_max=10 with a much smaller WD (e.g. 1e-4 or 1e-5)** to see if gentle regularization can stabilize the oscillation without causing divergence.

3. **AirfRANS needs a different approach** — the DrivAerML compound recipe doesn't transfer. The baseline lr=5e-4 + T_max=30 without gc/WD remains superior.

**Comment by tcapelle** [comment]
ADVISOR: Merging this PR. Run 3 (lr=6e-4 + gc=1.0 + T_max=10, no WD) achieves **4.431%** on DrivAerML — a 4.1% relative improvement over the 4.619% baseline. This is a clean result: the ablation structure clearly shows WD=1e-2 is destructive at lr=6e-4 (Runs 1 and 2 diverged), while removing WD yields a stable training run.

Key learnings:
- lr=6e-4 + gc=1.0 + T_max=10 is a strong DrivAerML config — better than the lr=5e-4 + T_max=30 baseline
- WD=1e-2 is incompatible with lr=6e-4 on DrivAerML (divergence)
- The DrivAerML recipe doesn't transfer to AirfRANS with these settings

New DrivAerML baseline: **4.431%** (this PR).
